### PR TITLE
DRILL-8279: Use thick Phoenix driver

### PR DESCRIPTION
**Pull request by vvysotskyi**
# [DRILL-8279](https://issues.apache.org/jira/browse/DRILL-8279): Use thick Phoenix driver

## Description

`phoenix-queryserver-client` shades Avatica classes, so it causes issues when starting Drill and shaded class from phoenix jars is loaded before official Avatica, so Drill wouldn't be able to start correctly.

- Fixed Phoenix unit tests to run with general tests suite without any extra checks
- Added support for Phoenix thick client
- Removed usage of custom dependencies

## Documentation
To use fat client, storage plugin configs should be updated as described in https://github.com/apache/drill/compare/master...vvysotskyi:DRILL-8279?expand=1#diff-81178bb7d4498bc22f7c54e71462b460f95b80f40f85e74ff8f1cc7f69d34351

## Testing
All unit tests pass, queried Phoenix using Drill.

### CODE REVIEW

**Comment by Z0ltrix** [comment]
Wow @vvysotskyi great work!
I'm so excited to see this in the real environment.
Regards

**Comment by jnturton** [comment]
@vvysotskyi even though we've only seen this Avatica conflict affect Drill master after the Calcite upgrade was merged, do I understand correctly that it also lies latent the 1.20 branch? If so then my thinking is that this change should be backported even though it represents a type of breaking change for existing users of PQS. Anyone on 1.20.x who needs to continue with PQS to the extent that they are willing to roll the classloader dice (I'd like to hope that this is no one) could still manually replace their Phoenix plugin with a download of drill-storage-phoenix-1.20.2.jar, the last version to be based on PQS, or just remain stuck on 1.20.2 until they can coordinate an upgrade...

**Review by jnturton — CHANGES_REQUESTED** [review]

File: contrib/storage-phoenix/pom.xml [hunk 1]
```diff
@@ -353,6 +381,9 @@
   <profiles>
     <profile>
       <id>hadoop-2</id>
+      <properties>
+        <phoenix.skip.tests>true</phoenix.skip.tests>
```
[the remark this answers is not in the document]
Comment (cgivre) [reply contrib/storage-phoenix/pom.xml:385]: See here: https://phoenix.apache.org/faq.html

```
Phoenix 4.x supports HBase 1.x running on Hadoop 2

Phoenix 5.x supports HBase 2.x running on Hadoop 3
```

[the remark this answers is not in the document]
Comment (vvysotskyi) [reply contrib/storage-phoenix/pom.xml:385]: For some reason CI tests fail with hadoop-2 profile, but pass fine locally. Seems like some issue with MiniKdc.

**Comment by cgivre** [comment]
Was there an agreement as to whether we should backport this to 1.20?

**Review by jnturton — APPROVED** [review]

**Comment by jnturton** [comment]
> Was there an agreement as to whether we should backport this to 1.20?

Not yet. For my part I'm supportive of backporting because I don't like the idea of a classloader time bomb ticking away in 1.20. As an alternative, 1.20 could remain on the thin driver but start distributing its own build of it with the needed Avatica relocations applied. But, since the thick driver is going to be preferred by most (all?) users, that seems like a wasted effort to me and I vote to backport this. Does anyone have a different take? @luocooong @vvysotskyi

**Comment by vvysotskyi** [comment]
I agree that this change should be backported, since classpath issue might affect any user, even if user doesn't use Phoenix at all.

**Comment by luocooong** [comment]
**_Holy cow, that's unbelievable !_**
Why did this get approved without production experience?
Why do we need to force use of fat clients?
There are ways to resolve package conflicts, but this pull request directly lets me discard using Drill to query Phoenix !

**Comment by jnturton** [comment]
@luocooong let's see if we can offer storage plugins for both thick and thin drivers then? The classpath conflict bug in the thin driver is very serious, even if you have not yet been affected. It was important that some action was taken immediately. 

Nothing has been released yet so we still have time to come with a path forward that works for everyone. Btw I thought I'd requested your review on this this PR as it was opened but I looking at the history I see must have failed to use the GH mobile app correctly so I do apologise if I only managed to bring this to your attention relatively late.

**Comment by vvysotskyi** [comment]
@luocooong, Drill plugins are still pluggable, so you can provide your own implementations if you need.

Phoenix official connectors [1] for such big data tools like Spark, Hive and Pig also use thick client, so such decision should be production-suitable. By the way I didn't find official connectors that use thin client in that repository.

Ideally, if Phoenix thin client shades some libraries, it should also relocate them to avoid such issues. I don't see any other correct way for resolving this class path conflict in other way. Creating specific module and repacking Phoenix there when building Drill is overhead, and didn't guarantee that nothing would be broken. Having custom repo that provides relocated classes also not good decision since it will make more complex supporting new Phoenix versions.

[1] https://github.com/apache/phoenix-connectors

**Comment by luocooong** [comment]
It was a real disappointment today.

As Java developers, all path conflicts will be solved in a way, but we chose one of the worst results.

What are we doing in such a hurry?

And I didn't see this pull request open for discussion before submission...

What does this mean for contributors?

**Comment by vvysotskyi** [comment]
@luocooong, at least I know several people that were already affected by this classpath issue, so it was the reason for fixing it quickly.

Jira ticket was created `14 Aug 14:36 EEST`, pull request was opened `14 Aug 17:53 EEST`, and merged `16 Aug 10:37 EEST`, so this time should be enough to participate in discussion or request some changes.

I didn't see a reason for creating a discussion in a mailing list for it, since plugin wasn't deleted, it is still functioning and even have improved developers experience by removing extra steps to run unit tests, enabling them in CI, and removing dependencies on custom repositories.

**Comment by luocooong** [comment]
Sorry, I cannot accept the above overbearing statement.
I don't want to argue any more about this pull request, because that pull request was actually merged, so the right to speak is your right.

**Comment by vvysotskyi** [comment]
@luocooong, sorry, I didn't aim to outspeak overbearing, I just wanted to express my thoughts on why I didn't start a discussion on the mailing list.

If you have ideas on how to avoid this classpath issue and use phoenix-queryserver-client, feel free to suggest them or create a pull request with changes.

**Pull request merged**